selfhost/codegen.jakt

```diff
@@ -0,0 +1,177 @@
+}
+
+struct ControlFlowState {
+    allowed_exits: AllowedControlExits
+    /// Whether `break` and `continue` should use `return ExplicitValueOrControlFlow` instead of
+    /// C++'s `break`/`continue`.
+    passes_through_match: bool
+    match_nest_level: usize
+
+    function no_control_flow() -> ControlFlowState {
+        return ControlFlowState(
+            allowed_exits: AllowedControlExits::Nothing
+            passes_through_match: false
+            match_nest_level: 0
+        )
+    }
+    function enter_function(this) -> ControlFlowState {
+        return ControlFlowState(
+            allowed_exits: AllowedControlExits::JustReturn
+            passes_through_match: false
+            match_nest_level: .match_nest_level
+        )
+    }
+    function enter_loop(this) -> ControlFlowState {
+        return ControlFlowState(
+            allowed_exits: AllowedControlExits::AtLoop
+            passes_through_match: false
+            match_nest_level: 0
+        )
+    }
+    function enter_match(this) -> ControlFlowState {
+        mut level = .match_nest_level
+        if .passes_through_match {
+            level = .match_nest_level + 1
+        }
+        return ControlFlowState(
+            allowed_exits: .allowed_exits.allow_return()
+            passes_through_match: true
+            match_nest_level: level
+        )
+    }
+    function is_match_nested(this) => .match_nest_level != 0
+    function choose_control_flow_macro(this) -> String {
+        if .allowed_exits.are_loop_exits_allowed() {
+            if .is_match_nested() {
+                return "JAKT_RESOLVE_EXPLICIT_VALUE_OR_CONTROL_FLOW_AT_LOOP_NESTED_MATCH"
+            }
+            return "JAKT_RESOLVE_EXPLICIT_VALUE_OR_CONTROL_FLOW_AT_LOOP"
+        }
+        return "JAKT_RESOLVE_EXPLICIT_VALUE_OR_CONTROL_FLOW_RETURN_ONLY"
+    }
+}
+
+struct CodeGenerator {
+
+    control_flow_state: ControlFlowState
+    entered_yieldable_blocks: [(String, String)] // label, variable name
+
+    function codegen_expression(mut this, expression: CheckedExpression) throws -> String {
+        return ""
+    }
+
+    function codegen_block(mut this, block: CheckedBlock) throws -> String {
+        mut output = ""
+
+        // FIXME: yielded_type
+
+        output += "{"
+
+        for statement in block.statements.iterator() {
+            output += .codegen_statement(statement)
+        }
+
+        output += "}"
+
+        // FIXME: yielded_type
+
+        return output
+    }
+
+    function codegen_statement(mut this, statement: CheckedStatement) throws -> String {
+        match statement {
+            Throw(expression) => {
+                return "return " + .codegen_expression(expression) + ";"
+            }   
+            Continue => {
+                if .control_flow_state.passes_through_match {
+                    return "return JaktInternal::LoopContinue{};"
+                }
+                return "continue;"
+            }
+            Continue => {
```

Comment:
You've got `Continue` twice

Reply:
this should definitely make the typechecker get mad at you, how did I manage to miss making this an error...?

Reply:
That I do!